### PR TITLE
WIP Simplify and extend 2D torch tests

### PR DESCRIPTION
**Pull request by janden**
We reparametrize the tests using backend–device tuples, extend tests to increase coverage, and make some smaller cosmetic changes. Ideally, this should simplify writing the tests for the TF frontend.

### CODE REVIEW

**Comment by lostanlen** [comment]
This `.to()` method is really coming in handy!

File: kymatio/scattering2d/tests/test_torch_scattering2d.py [hunk 1]
```diff
@@ -7,66 +7,123 @@
...
     if torch.cuda.is_available():
         from skcuda import cublas
         import cupy
         from kymatio.scattering2d.backend.torch_skcuda_backend import backend
         backends.append(backend)
+        if 'cuda' in devices:
+            backends_devices.append((backend, 'cuda'))
 except:
-    pass
+    Warning('torch_skcuda backend not available.')
```
Comment (edouardoyallon) [inline kymatio/scattering2d/tests/test_torch_scattering2d.py]: Alright this sounds better.

File: kymatio/scattering2d/tests/test_torch_scattering2d.py [hunk 2]
```diff
@@ -7,66 +7,123 @@
...
+            modulus(y)
+        assert 'should be complex' in record.value.args[0]
+
         y = x[::2, ::2]
         with pytest.raises(RuntimeError) as record:
             modulus(y)
         assert 'should be contiguous' in record.value.args[0]
 
+    @pytest.mark.parametrize('backend', backends)
+    def test_cuda_only(self, backend):
```
Comment (edouardoyallon) [inline kymatio/scattering2d/tests/test_torch_scattering2d.py]: ok

File: kymatio/scattering2d/tests/test_torch_scattering2d.py [hunk 3]
```diff
@@ -81,20 +138,33 @@ def test_SubsampleFourier(self, device, backend):
...
             subsample_fourier(y, k=16)
         assert 'should be contiguous' in record.value.args[0]
 
-    @pytest.mark.parametrize("device", devices)
-    @pytest.mark.parametrize("backend", backends)
-    def test_batch_shape_agnostic(self, device, backend):
-        if device == 'cpu' and backend.name == 'torch_skcuda':
-            return
+    @pytest.mark.parametrize('backend', backends)
+    def test_gpu_only(self, backend):
```
Comment (edouardoyallon) [inline kymatio/scattering2d/tests/test_torch_scattering2d.py]: ok

File: kymatio/scattering2d/tests/test_torch_scattering2d.py (hunk 1 above)
Comment (janden) [reply kymatio/scattering2d/tests/test_torch_scattering2d.py]: Yes and it has the advantage of getting us to :100: test coverage... (For some reason, `pass` always counts as a miss for coverage purposes.)

File: kymatio/scattering2d/tests/test_torch_scattering2d.py [hunk 4]
```diff
@@ -227,43 +392,58 @@ def test_Scattering2D(self, backend, device):
         M = x.shape[2]
         N = x.shape[3]
 
-        # Then, let's check when using pure pytorch code
-        scattering = Scattering2D(J, shape=(M, N), pre_pad=pre_pad, backend=backend, frontend='torch')
+        scattering = Scattering2D(J, shape=(M, N), pre_pad=pre_pad,
+                                  backend=backend, frontend='torch')
```
Comment (edouardoyallon) [inline kymatio/scattering2d/tests/test_torch_scattering2d.py]: remove the newline?

Comment (janden) [reply kymatio/scattering2d/tests/test_torch_scattering2d.py]: Wanted to keep lines below 80 characters.

File: kymatio/scattering2d/tests/test_torch_scattering2d.py [hunk 5]
```diff
@@ -227,43 +392,58 @@ def test_Scattering2D(self, backend, device):
...
+                                  backend=backend)
+
+        S1x = scattering(x)
+        assert torch.allclose(S1x, S[..., 0:len(o0 + o1), :, :])
+
+    @pytest.mark.parametrize('backend', backends)
+    def test_gpu_only(self, backend):
+        if backend.name == 'torch_skcuda':
+            scattering = Scattering2D(3, shape=(32, 32), backend=backend,
+                                      frontend='torch')
```
Comment (edouardoyallon) [inline kymatio/scattering2d/tests/test_torch_scattering2d.py]: new line?

Comment (janden) [reply kymatio/scattering2d/tests/test_torch_scattering2d.py]: See above.

File: kymatio/scattering2d/tests/test_torch_scattering2d.py (hunk 4 above)
Comment (edouardoyallon) [reply kymatio/scattering2d/tests/test_torch_scattering2d.py]: OK.

File: kymatio/scattering2d/tests/test_torch_scattering2d.py (hunk 5 above)
Comment (edouardoyallon) [reply kymatio/scattering2d/tests/test_torch_scattering2d.py]: OK.

**Comment by janden** [comment]
> Ok this is a great great simplification.

Great. Unfortunately, some of the tests seem to be bugging out after the rebase. Will fix and then merge.

**Pull request merged**